### PR TITLE
[FLOC-4194] Add GCE functional test that verifies the attached to other node exception.

### PR DESCRIPTION
**Pull request by sarum90**
Refactors the provisioner so that it can be used by code that does not need to log into the instance. Adds a new test to the GCE functional tests that actually spins up a second instance and attaches a volume to that instance just to verify the exception raised by the BlockDeviceAPI.

There are other tests that will also use this infrastructure forthcoming.

### CODE REVIEW

File: flocker/node/agents/gce.py [hunk 1]
```diff
@@ -200,6 +200,31 @@ def _get_metadata_path(path):
     return r.text
 
 
+def get_machine_zone():
+    """
+    Returns the zone that the GCE instance running this code is running within.
+
+    It does this by querying the metadata server.
+
+    :returns unicode: The zone where the curring instance is running.
```
Comment (zubron) [inline flocker/node/agents/gce.py]: `current` rather than `curring`?

File: flocker/node/agents/gce.py [hunk 2]
```diff
@@ -200,6 +200,31 @@ def _get_metadata_path(path):
...
+
+
+def get_machine_project():
+    """
+    Returns the project that the GCE instance running this code is running
+    within.
+
+    It does this by querying the metadata server.
+
+    :returns unicode: The project under which the curring instance is running.
```
Comment (zubron) [inline flocker/node/agents/gce.py]: Same as above.

File: flocker/provision/_gce.py [hunk 3]
```diff
@@ -272,32 +297,71 @@ def _create_gce_instance_config(instance_name, project, zone, machine_type,
...
+        with start_action(
+            action_type=u"flocker:provision:gce:destroy",
+            instance_id=self.name,
+        ):
+            operation = self.compute.instances().delete(
+                project=self.project,
+                zone=self.zone,
+                instance=self.name
+            ).execute()
+            wait_for_operation(self.compute, operation, timeout_steps=[5]*60)
```
Comment (zubron) [inline flocker/provision/_gce.py]: :question: Any particular reason for changing the timeout values?

File: flocker/provision/_gce.py [hunk 4]
```diff
@@ -364,25 +419,86 @@ def reboot(self):
...
 
+    def create_instance(self, instance_name, image=None, **kwargs):
+        """
+        Create a GCE instance.
+
+        :param unicode instance: The name of the instance to create.
+        :param unicode image: The link to the GCE image to use for the base
+            disk. Defaults to the latest CentOS 7 image if unspecified.
+        :param **kwargs: Other keyword arguments for the creation of the
+            instance see documentation of ``_create_gce_instance_config``
```
Comment (zubron) [inline flocker/provision/_gce.py]: Use two separate sentences here. `...instance. See documentation...`

**Comment by zubron** [comment]
Thanks @sarum90. This looks good. Please address the comments and merge when the tests are green.

File: flocker/node/agents/gce.py (hunk 1 above)
Comment (sarum90) [inline flocker/node/agents/gce.py]: Thanks, fixed.

File: flocker/node/agents/gce.py (hunk 2 above)
Comment (sarum90) [inline flocker/node/agents/gce.py]: Thanks. I even had spell check on... but alas: http://dictionary.reference.com/browse/curring

File: flocker/provision/_gce.py (hunk 3 above)
Comment (sarum90) [inline flocker/provision/_gce.py]: Nice catch, you're keeping me honest :).

Yea I did see it take somewhere between 3 and 4 minutes once when I ran the test.

As this is test infra code, the right answer seemed to be just wait up to 5 minutes. I also figured if it was going to take that long we probably don't need to poll every second.

Added a comment so people don't think about making it shorter.

File: flocker/provision/_gce.py (hunk 4 above)
Comment (sarum90) [inline flocker/provision/_gce.py]: Done.

**Comment by sarum90** [comment]
Thanks for the review @zubron! I realized I should run the GCE acceptance tests as well to verify I didn't break them, so I'm going to do that before I merge. Otherwise the changes are all comment changes, so I might just wait for run_lint.

**Pull request merged**